### ISSUE TITLE
Subs in Subs returns incorrect results

### ISSUE THREAD

**Issue by certik** [sympy/sympy#8768]
First without cache:
```
certik@redhawk:~/repos/sympy((7d7eb67...))$ bin/isympy -C
IPython console for SymPy 0.7.6-git (Python 2.7.6-64-bit) (ground types: python, cache: off)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at http://docs.sympy.org/dev

In [1]: f(x).diff(x).subs(x, 0)
Out[1]: 
⎛d       ⎞│   
⎜──(f(x))⎟│   
⎝dx      ⎠│x=0

In [2]: _.subs(x, y)
Out[2]: 
⎛d       ⎞│   
⎜──(f(x))⎟│   
⎝dx      ⎠│x=y
```
However, the output of [2] should be:
```
⎛d       ⎞│   
⎜──(f(x))⎟│   
⎝dx      ⎠│x=0
```
because [1] does not have any `x` in the final expression (it is substituted for 0). This is a major bug.

With caching enabled, there are more bugs, we found for example:
```
certik@redhawk:~/repos/sympy((7d7eb67...))$ ./bin/isympy 
IPython console for SymPy 0.7.6-git (Python 2.7.8-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at http://docs.sympy.org/dev

In [1]: f(x).diff(x, x).subs(x, 0)
Out[1]: 
⎛  2      ⎞│   
⎜ d       ⎟│   
⎜───(f(x))⎟│   
⎜  2      ⎟│   
⎝dx       ⎠│x=0

In [2]: var("h")
Out[2]: h

In [3]: f(x).diff(x, x).subs(x, 0)
Out[3]: 
⎛  2      ⎞│   
⎜ d       ⎟│   
⎜───(f(x))⎟│   
⎜  2      ⎟│   
⎝dx       ⎠│x=0

In [4]: 

In [4]: _.subs(x, h)
Out[4]: 
⎛  2      ⎞│   
⎜ d       ⎟│   
⎜───(f(x))⎟│   
⎜  2      ⎟│   
⎝dx       ⎠│x=h

In [5]: f(x).diff(x, x).subs(x, 0)
Out[5]: 
⎛  2      ⎞│   
⎜ d       ⎟│   
⎜───(f(x))⎟│   
⎜  2      ⎟│   
⎝dx       ⎠│x=0

In [6]: _.subs(x, h)
Out[6]: 
⎛  2      ⎞│   
⎜ d       ⎟│   
⎜───(f(x))⎟│   
⎜  2      ⎟│   
⎝dx       ⎠│x=h

In [7]: f(x).diff(x).subs(x, 0)
Out[7]: 
⎛d       ⎞│   
⎜──(f(x))⎟│   
⎝dx      ⎠│x=0

In [8]: _.subs(x, h)
Out[8]: 
⎛d       ⎞│   
⎜──(f(x))⎟│   
⎝dx      ⎠│x=h

In [9]: x*(f(x).diff(x).subs(x, 0))
Out[9]: 
  ⎛d       ⎞│   
x⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=0

In [10]: (x*(f(x).diff(x).subs(x, 0))).subs(x, h)
Out[10]: 
  ⎛d       ⎞│   
h⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=h

In [11]: Subs?

In [12]: Subs??

In [13]: (x*(f(x).diff(x).subs(x, 0))).subs(x, h)
Out[13]: 
  ⎛d       ⎞│   
h⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=h

In [14]: (x*(f(x).subs(x, y).diff(y).subs(y, 0))).subs(x, h)
Out[14]: 
  ⎛d       ⎞│   
h⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=h

In [15]: (x*(f(x).subs(x, y).diff(y).subs(y, 0)))
Out[15]: 
  ⎛d       ⎞│   
x⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=0

In [16]: (x*(f(x).subs(x, y).diff(y).subs(y, 0)))
Out[16]: 
  ⎛d       ⎞│   
x⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=0

In [17]: f(x).subs(x, y)
Out[17]: f(y)

In [18]: f(x).subs(x, y).diff(y)
Out[18]: 
d       
──(f(y))
dy      

In [19]: f(x).subs(x, y).diff(y).subs(y, 0)
Out[19]: 
⎛d       ⎞│   
⎜──(f(y))⎟│   
⎝dy      ⎠│y=0

In [20]: x*(f(x).subs(x, y).diff(y).subs(y, 0))
Out[20]: 
  ⎛d       ⎞│   
x⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=0

In [21]: f(x).subs(x, y).diff(y).subs(y, 0)
Out[21]: 
⎛d       ⎞│   
⎜──(f(y))⎟│   
⎝dy      ⎠│y=0

In [22]: v = f(x).subs(x, y).diff(y).subs(y, 0)

In [23]: v
Out[23]: 
⎛d       ⎞│   
⎜──(f(y))⎟│   
⎝dy      ⎠│y=0

In [24]: x*v
Out[24]: 
  ⎛d       ⎞│   
x⋅⎜──(f(x))⎟│   
  ⎝dx      ⎠│x=0
```
I.e. in [23] the `v` just contains `y`, but by multiplying by `x` in [24], suddenly `y` is changed for `x`... This test case can probably be simplified, as this behavior depends on the prompts [1] - [21]. If you only execute [22], [23] and [24] in a new shell, then it works.

@pbrady and I found this together.

**Comment by certik** [comment]
The fix is simple:
```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 43d22e6..1935085 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -1557,7 +1557,7 @@ def _hashable_content(self):
     def _eval_subs(self, old, new):
         if old in self.variables:
             pts = list(self.point.args)
-            pts[self.variables.index(old)] = new
+            #pts[self.variables.index(old)] = new
             return self.func(self.expr, self.variables, pts)
 
     def _eval_derivative(self, s):
```
I.e. the code in `Subs._eval_subs()` seems to be completely wrong. With this patch, only one test fails in the core:
```
____________________________________ sympy/core/tests/test_function.py:test_Subs ____________________________________
  File "/home/certik/repos/sympy/sympy/core/tests/test_function.py", line 186, in test_Subs
    assert Subs(x, x, 0).subs(x, 1) == Subs(x, x, 1)
AssertionError

 tests finished: 1469 passed, 1 failed, 77 skipped, 201 expected to fail, 1 expected to fail but passed, 
in 146.65 seconds 
```
and this test seems to be completely wrong, i.e. on the LHS Subs(x, x, 0) is simply 0, so if you do 0.subs(x, 1), the answer is 0. But the RHS is 1, so these two are not equal.

**Comment by certik** [comment]
Proper fix and tests submitted in #8769.

**Comment by pbrady** [comment]
It's interesting that there is a test for this behavior.  I'm assuming this means at some point that someone thought about this and decided it was a good idea.  I suppose it does have the benefit that `Subs` is treated like any other expression by `subs` whereas we are saying it should be treated differently.
